src/core/thread/router_table.cpp

```diff
@@ -0,0 +1,634 @@
+        {
+            router = &mRouters[i];
+            ExitNow();
+        }
+    }
+
+exit:
+    return router;
+}
+
+Router *RouterTable::GetRouter(uint8_t aRouterId)
+{
+    Router * rval   = NULL;
+    uint16_t rloc16 = Mle::Mle::GetRloc16(aRouterId);
+
+    for (uint8_t i = 0; i < Mle::kMaxRouters; i++)
+    {
+        if (mRouters[i].GetRloc16() == rloc16)
+        {
+            rval = &mRouters[i];
+            ExitNow();
+        }
+    }
+
+exit:
+    return rval;
+}
+
+const Router *RouterTable::GetRouter(uint8_t aRouterId) const
+{
+    const Router *rval   = NULL;
+    uint16_t      rloc16 = Mle::Mle::GetRloc16(aRouterId);
+
+    for (uint8_t i = 0; i < Mle::kMaxRouters; i++)
+    {
+        if (mRouters[i].GetRloc16() == rloc16)
+        {
+            rval = &mRouters[i];
+            ExitNow();
+        }
+    }
+
+exit:
+    return rval;
+}
+
+Router *RouterTable::GetRouter(const Mac::ExtAddress &aExtAddress)
+{
+    Router *router = NULL;
+
+    for (int i = 0; i < Mle::kMaxRouters; i++)
+    {
+        if (mRouters[i].GetRloc16() == 0xffff)
+        {
+            break;
+        }
+
+        if (mRouters[i].GetExtAddress() == aExtAddress)
+        {
+            router = &mRouters[i];
+            ExitNow();
+        }
+    }
+
+exit:
+    return router;
+}
+
+otError RouterTable::GetRouterInfo(uint16_t aRouterId, otRouterInfo &aRouterInfo)
+{
+    otError error = OT_ERROR_NONE;
+    Router *router;
+    uint8_t routerId;
+
+    if (aRouterId <= Mle::kMaxRouterId)
+    {
+        routerId = static_cast<uint8_t>(aRouterId);
+    }
+    else
+    {
+        VerifyOrExit(Mle::Mle::IsActiveRouter(aRouterId), error = OT_ERROR_INVALID_ARGS);
+        routerId = Mle::Mle::GetRouterId(aRouterId);
+        VerifyOrExit(routerId <= Mle::kMaxRouterId, error = OT_ERROR_INVALID_ARGS);
+    }
+
+    memset(&aRouterInfo, 0, sizeof(aRouterInfo));
+    aRouterInfo.mRouterId = routerId;
+    aRouterInfo.mRloc16   = Mle::Mle::GetRloc16(routerId);
+
+    router = GetRouter(routerId);
+
+    if (router != NULL)
+    {
+        aRouterInfo.mExtAddress      = router->GetExtAddress();
+        aRouterInfo.mAllocated       = true;
+        aRouterInfo.mNextHop         = router->GetNextHop();
+        aRouterInfo.mLinkEstablished = router->GetState() == Neighbor::kStateValid;
+        aRouterInfo.mPathCost        = router->GetCost();
+        aRouterInfo.mLinkQualityIn   = router->GetLinkInfo().GetLinkQuality();
+        aRouterInfo.mLinkQualityOut  = router->GetLinkQualityOut();
+        aRouterInfo.mAge = static_cast<uint8_t>(TimerMilli::MsecToSec(TimerMilli::GetNow() - router->GetLastHeard()));
+    }
+
+exit:
+    return error;
+}
+
+Router *RouterTable::GetLeader(void)
+{
+    return GetRouter(GetNetif().GetMle().GetLeaderId());
+}
+
+uint32_t RouterTable::GetLeaderAge(void) const
+{
+    return TimerMilli::MsecToSec(TimerMilli::GetNow() - mRouterIdSequenceLastUpdated);
+}
+
+uint8_t RouterTable::GetActiveRouterCount(void) const
+{
```

Comment:
How about defining a member variable to save the active router counter which could be determined when build the index map in `UpdateAllocation()`?